archive/import.go

```diff
@@ -2,156 +2,350 @@ package archive
+// footprint.
+func (iw *importWriter) spillLargestBuffer() error {
+	var dw *tsDirWriter
+	for _, w := range iw.writers {
+		if dw == nil || w.size > dw.size {
+			dw = w
+		}
+	}
+	return dw.spill()
+}
+
+func (iw *importWriter) close() error {
+	var merr error
+	for ts, w := range iw.writers {
+		if err := w.flush(); err != nil {
+			merr = multierr.Append(merr, err)
+		}
+		delete(iw.writers, ts)
+	}
+	return merr
+}
+
+// tsDirWriter accumulates records for one tsDir.
+// When the expected size of writing the records is greater than
+// ark.LogSizeThreshold, they are written to a chunk file in
+// the archive.
+type tsDirWriter struct {
+	ark          *Archive
+	ctx          context.Context
+	importWriter *importWriter
+	records      []*zng.Record
+	size         int64
+	spiller      *spill.MergeSort
+	tsDir        tsDir
+	uri          iosrc.URI
+}
+
+func newTsDirWriter(iw *importWriter, tsDir tsDir) (*tsDirWriter, error) {
+	d := &tsDirWriter{
+		ark:          iw.ark,
+		ctx:          iw.ctx,
+		importWriter: iw,
+		tsDir:        tsDir,
+		uri:          iw.ark.DataPath.AppendPath(dataDirname, tsDir.name()),
+	}
+	if dirmkr, ok := d.ark.dataSrc.(iosrc.DirMaker); ok {
+		if err := dirmkr.MkdirAll(d.uri, 0755); err != nil {
+			return nil, err
+		}
+	}
+	return d, nil
+}
+
+func (dw *tsDirWriter) addBufSize(delta int64) {
+	old := dw.bufSize()
+	dw.size += delta
+	dw.importWriter.memBuffered += dw.bufSize() - old
+}
+
+// bufSize returns the actually buffer size as a multiple of importLZ4BlockSize.
+// This is done to ensure that whenever spill the generated spill is compressed
+// nicely onto disk.
+func (dw *tsDirWriter) bufSize() int64 {
+	return (dw.size / int64(importLZ4BlockSize)) * int64(importLZ4BlockSize)
+}
+
+// approxTotalBytes is the sum of the size of compressed records spilt to disk
+// and a crude approximation of the buffer record bytes (simply bufBytes / 2).
+func (dw *tsDirWriter) approxTotalBytes() int64 {
+	b := (dw.bufSize() / 2)
+	if dw.spiller != nil {
+		b += dw.spiller.SpillSize()
+	}
+	return b
+}
+
+func (dw *tsDirWriter) writeOne(rec *zng.Record) error {
+	dw.records = append(dw.records, rec)
+	dw.addBufSize(int64(len(rec.Raw)))
+	if dw.approxTotalBytes() > dw.ark.LogSizeThreshold {
+		if err := dw.flush(); err != nil {
+			return err
+		}
 	}
 	return nil
 }
 
-func (d *importDriver) writeOne(rec *zng.Record) error {
-	if d.dataFileWriter != nil && !d.firstTs.DayOf().Contains(rec.Ts()) {
-		// Don't allow data files to include records from multiple tsDir time spans.
-		if err := d.close(); err != nil {
+func (dw *tsDirWriter) spill() error {
+	if len(dw.records) == 0 {
+		return nil
+	}
+	if dw.spiller == nil {
+		var err error
+		dw.spiller, err = spill.NewMergeSort(importCompareFn(dw.ark))
+		if err != nil {
 			return err
 		}
+		dw.spiller.SetWriterOpts(zngio.WriterOpts{
+			LZ4BlockSize:     importLZ4BlockSize,
+			StreamRecordsMax: importStreamRecordsMax,
+		})
```

Comment:
Based on https://github.com/brimsec/zq/pull/1050#discussion_r463670238, we probably don't want to compress these files, and there's no benefit to giving them a nonzero StreamRecordsMax, so this and the zngio.WriterOpts additions in proc/spill can go away.

Reply:
Sure, I'll remove this.